tapestry.js

```diff
@@ -48,18 +51,50 @@ jQuery.get(apiUrl + "/tapestries/" + tapestryWpPostId, function(result){
 
     tapestrySlug = dataset.settings.tapestrySlug;
     
-    if (saveProgressToCookie) {
-        // Update dataset with data from cookie (if any)
-        var cookieProgress = Cookies.get("progress-data-"+tapestrySlug);
-        if (cookieProgress !== undefined) {
-            cookieProgress = JSON.parse( cookieProgress );
-            setDatasetProgress(cookieProgress);
-        }
-        // Update H5P Video Settings from cookie (if any)
-        var cookieH5PVideoSettings = Cookies.get("h5p-video-settings");
-        if (cookieH5PVideoSettings !== undefined) {
-            cookieH5PVideoSettings = JSON.parse( cookieH5PVideoSettings );
-            h5pVideoSettings = cookieH5PVideoSettings;
+    if (saveProgress) {
+        // If user is logged in, get progress from database database
+        if (tapestryWpUserId) {
+            var getProgData = {
+                "post_id": tapestryWpPostId,
+            };
+
+            jQuery.get(TAPESTRY_PROGRESS_URL, getProgData, function(result) {
+                if (result && !isEmptyObject(result)) {
+                    setDatasetProgress(JSON.parse(result));
+                    updateViewedProgress(); // update viewed progress because async fetch of dataset
+                }
+            }).fail(function(e) {
+                console.log("Error with retrieving node progress");
+                console.log(e);
+            });
+
+            var getH5PData = {
+                "post_id": tapestryWpPostId
+            };
+            jQuery.get(TAPESTRY_H5P_SETTINGS_URL, getH5PData, function(result) {
+                if (result && !isEmptyObject(result)) {
+                    h5pVideoSettings = JSON.parse(result);
+                }
+            }).fail(function(e) {
+                console.log("Error with retrieving h5p video settings");
+                console.log(e);
+            });
+
+        } else { 
+            // Update dataset with data from cookie (if any)
+            var cookieProgress = Cookies.get("progress-data-"+tapestrySlug);
+
+            if (cookieProgress) {
+                cookieProgress = JSON.parse( cookieProgress );
+                setDatasetProgress(cookieProgress);	
```

Comment:
Don't you also need to call `updateViewedProgress();` like above here as well?

Reply:
It's not needed for the cookie progress since it is synchronous and the updating of the progress will be done after the dataset is set in the buildNodeContents function somewhere. It's needed for the database request since it's async.